sdk/iot/core/inc/az_iot_provisioning_client.h

```diff
@@ -0,0 +1,310 @@
+ * @param signature An empty #az_span with sufficient capacity to hold the SAS signature.
+ * @param out_signature The output #az_span containing the SAS signature.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_provisioning_client_sas_signature_get(
+    az_iot_provisioning_client const* client,
+    uint32_t token_expiration_unix_time,
+    az_span signature,
+    az_span* out_signature);
+
+/**
+ * @brief Gets the MQTT password.
+ * @note The MQTT password must be an empty string if X509 Client certificates are used. Use this
+ *       API only when authenticating with SAS tokens.
+ *
+ * @param client
+ * @param base64_hmac_sha256_signature The Base64 encoded value of the HMAC-SHA256(signature,
+ *                                     SharedAccessKey). The signature is obtained by using
+ *                                     #az_iot_hub_client_sas_signature_get.
+ *
+ * @param key_name The Shared Access Key Name (Policy Name). This is optional. For security reasons
+ *                 we recommend using one key per device instead of using a global policy key.
+ * @param mqtt_password An empty #az_span with sufficient capacity to hold the MQTT password.
+ * @param out_mqtt_password The output #az_span containing the MQTT password.
+ * @return #az_result.
+ */
+AZ_NODISCARD az_result az_iot_hub_client_sas_password_get(
+    az_iot_provisioning_client const* client,
+    az_span base64_hmac_sha256_signature,
+    az_span key_name,
+    az_span mqtt_password,
+    az_span* out_mqtt_password);
+
+/**
+ *
+ * Register APIs
+ *
+ *   Use the following APIs when the Shared Access Key is available to the application or stored
+ *   within a Hardware Security Module. The APIs are not necessary if X509 Client Certificate
+ *   Authentication is used.
+ */
+
+/**
+ * @brief Azure IoT Provisioning status codes.
+ *
+ */
+typedef enum
+{
+  // Service success codes
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_OK = 200,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_ACCEPTED = 202,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_NO_CONTENT = 204,
+
+  // Service error codes
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_BAD_REQUEST = 400,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_UNAUTHORIZED = 401,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_FORBIDDEN = 403,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_NOT_FOUND = 404,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_NOT_ALLOWED = 405,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_NOT_CONFLICT = 409,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_PRECONDITION_FAILED = 412,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_REQUEST_TOO_LARGE = 413,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_UNSUPPORTED_TYPE = 415,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_THROTTLED = 429,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_CLIENT_CLOSED = 499,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_SERVER_ERROR = 500,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_BAD_GATEWAY = 502,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_SERVICE_UNAVAILABLE = 503,
+  AZ_IOT_PROVISIONING_CLIENT_STATUS_TIMEOUT = 504,
+} az_iot_provisioning_client_status;

```

Comment:
In az_http.h, we have az_http_status_code, do you want to use that?

Reply:
We brought that up in the review with @JeffreyRichter and the decision was that we want to separate the status codes. These are IoT Hub / Provisioning specific and come over proprietary protocol over MQTT. (I don't have any data about this but it feels better to consider the values subject to change in future versions of our services.)

Reply:
Yes, these should be separate; it is an implementation coincidence that they match HTTP status codes. Also, it would not look right to compare status codes returned via MQTT with az_http_status_code.